csharp/src/Ice/UriParser.cs

```diff
@@ -0,0 +1,340 @@
+                            uri = InitialParse(altUriString, generalOptions: null, endpointOptions);
+
+                            // > 1 because there is always a first empty segment.
+                            if (uri.AbsolutePath.Length > 1 || uri.Fragment.Length > 0)
+                            {
+                                throw new FormatException($"alt-endpoint `{alt}' must not specify a path or fragment");
+                            }
+                            endpoints.Add(CreateEndpoint(communicator,
+                                                         oaEndpoints,
+                                                         endpointOptions,
+                                                         protocol,
+                                                         uri,
+                                                         alt));
+                        }
+                    }
+                }
+
+                return (encoding,
+                        (IReadOnlyList<Endpoint>?)endpoints ?? (IReadOnlyList<Endpoint>)Array.Empty<Endpoint>(),
+                        facet,
+                        invocationMode,
+                        path,
+                        protocol);
+            }
+            catch (Exception ex)
+            {
+                // Give context to the exception.
+                throw new FormatException($"failed to parse URI `{uriString}'", ex);
+            }
+        }
+
+        /// <summary>Registers the ice and ice+universal schemes.</summary>
+        internal static void RegisterCommon()
+        {
+            RegisterTransport("universal", defaultPort: 0);
+
+            // There is actually no authority at all with the ice scheme, but we emulate it with an empty authority
+            // during parsing by the Uri class and the GenericUriParser.
+            GenericUriParserOptions options =
+                ParserOptions |
+                GenericUriParserOptions.AllowEmptyAuthority |
+                GenericUriParserOptions.NoPort;
+
+            System.UriParser.Register(new GenericUriParser(options), "ice", -1);
+        }
+
+        /// <summary>Registers an ice+transport scheme.</summary>
+        /// <param name="transportName">The name of the transport (cannot be empty).</param>
+        /// <param name="defaultPort">The default port for this transport.</param>
+        internal static void RegisterTransport(string transportName, ushort defaultPort) =>
+            System.UriParser.Register(new GenericUriParser(ParserOptions), $"ice+{transportName}", defaultPort);
+
+        private static Endpoint CreateEndpoint(
+            Communicator communicator,
+            bool oaEndpoint,
+            Dictionary<string, string> options,
+            Protocol protocol,
+            Uri uri,
+            string uriString)
+        {
+            string transportName = uri.Scheme.Substring(4); // i.e. chop-off "ice+"
```

Comment:
We can assert this is only call for `ice+` scheme
```suggestion
           Debug.Assert(uri.Scheme.StartsWith("ice+"));
            string transportName = uri.Scheme.Substring(4); // i.e. chop-off "ice+"
```

Reply:
Added assert as suggested.